File: Python/creating_maps.py
import plotly.express as px
import win32com.client
import pandas as pd
c = win32com.client.Dispatch("Macrobond.Connection")
d = c.Database

#Getting the region key
s = d.FetchOneSeries("owidvacci_ar_pfvph")
key = s.Metadata.GetFirstValue("RegionKey") #Getting the region key of a series

regionkey = d.FetchOneEntity(key) 
regionkeydesc = regionkey.Metadata.GetFirstValue("Description") #Getting the full name of the concept

regionkeydesc

#Searching for all series with the region key/concept 
query = d.CreateSearchQuery()
query.SetEntityTypeFilter("TimeSeries")
query.AddAttributeValueFilter("RegionKey", key)
result = d.Search(query).Entities

#Getting the last value and the region of the series via the metadata
val = pd.DataFrame(columns=['value'])
region = pd.DataFrame(columns=['region'])

for s in result:
    region= region.append({'region': s.Metadata.GetFirstValue("Region")}, ignore_index=True)
    val= val.append({'value': s.Metadata.GetFirstValue("LastValue")}, ignore_index=True)

#Translating the region code to ISO 3 and accessing the full name of the country
li=region.values.tolist()

iso = pd.DataFrame(columns=['iso'])
country = pd.DataFrame(columns=['country'])

entities = d.FetchEntities(li)

for e in entities:
    iso= iso.append({'iso': e.Metadata.GetFirstValue("IsoCountryCode3")}, ignore_index=True)
    country= country.append({'country': e.Metadata.GetFirstValue("Description")}, ignore_index=True)

#Merging the ISO3, Country name, Region and Values into one 
df = pd.merge(iso, country, left_index=True, right_index=True)
df=pd.merge(df, region, left_index=True, right_index=True)
df=pd.merge(df, val, left_index=True, right_index=True)
df['iso'] = df['iso'].str.upper()

import plotly.graph_objects as go

fig = go.Figure(data=go.Choropleth(
    locations = df['iso'],
    z = df['value'],
    text = df['country'],
    colorscale = 'viridis',
    autocolorscale=False,
    reversescale=True,
    marker_line_color='darkgray',
    marker_line_width=0.5,
    colorbar_title = '',
))

# Adding a dynamic title and sourcing
fig.update_layout(
    title_text=regionkeydesc,
    geo=dict(
        showframe=False,
        showcoastlines=False,
        projection_type='equirectangular'
    ),
    annotations = [dict(
        x=0.55,
        y=0.00001,
        xref='paper',
        yref='paper',
        text='Source: <a href="https://www.macrobond.com/">Macrobond</a>',
        showarrow = False
    )]
)

File: R/creating_maps.r
library(MacrobondAPI) #This sample requires Macrobond API 1.2-4 or later

library(ggplot2)

library(tidyverse)

library(rnaturalearth)

library(rnaturalearthdata)

library(plotly)

GetRegion <- function(entity) getFirstMetadataValue(getMetadata(entity), "Region")

GetRegionNames <- function(entities) sapply(entities, GetRegion)

GetLastValueFromMetadata <- function(entity) getFirstMetadataValue(getMetadata(entity), "LastValue")

GetRegionsIso2 <- function(regionEntities) sapply(regionEntities, function(entity) getFirstMetadataValue(getMetadata(entity), "IsoCountryCode2"))

FetchOneTimeSeries()

‍

# Searching for a concept in Macrobond and accessing the available data globally

query <- CreateSearchQuery()

setEntityTypeFilter(query, "TimeSeries")

addAttributeValueFilter(query, "RegionKey" , "owdc0008")

searchResult <- SearchEntities(query)

entities <- getEntities(searchResult)

regionNames <- GetRegionNames(entities) #Getting the Region names from the search result

regionEntities <- FetchEntities(regionNames)#Getting the Region Entities from Region in order to translate it to ISO2

regionIso <- GetRegionsIso2(regionEntities) #Getting the ISO 2

entitiesAndRegions <- mapply(function (entity, iso) c(entity = entity, iso = iso), entities, regionIso) #Converting the ISO

‍

# Inserting the country list for the map from rnaturalearth

mapdata <- ne_countries(scale = "medium", returnclass = "sf")

‍

# Remove entries for which we do not have matching iso codes

entitiesAndRegions <- entitiesAndRegions[unlist(sapply(entitiesAndRegions, function(x) !is.null(x$iso) && x$iso %in% mapdata$iso_a2))]

lastValues <- sapply(entitiesAndRegions, function(x) GetLastValueFromMetadata(x$entity))#Getting the last value from the meta data

isoCodes <- sapply(entitiesAndRegions, function(x) x$iso) #Getting the ISO2 codes

‍

# Creating a dataframe with region and last value

insert_data <- data.frame(iso_a2 = isoCodes, value = lastValues)

‍

# Joining the values with the map dataset

mapdata <- left_join(mapdata, insert_data, by="iso_a2")

mapdata <- subset(mapdata, select = c(name, value, geometry, iso_a2))

library(viridis)

‍

# Creating the map

map <- ggplot(data = mapdata, aes(text = paste(name, "<br>", "Fully vaccinated:", value,"%"))) +

geom_sf(aes(fill = value))+

ggtitle(GetConceptDescription(getConcepts(entities[[1]])))+

labs(title = GetConceptDescription(getConcepts(entities[[1]])), x = "Source: Macrobond")+

scale_fill_viridis_c(name = "%", option = "viridis", trans= "reverse")+

theme(axis.text.x = element_blank(),

axis.text.y = element_blank(),

axis.ticks = element_blank(),

axis.title.y=element_blank(),

rect = element_blank())

map

‍

# Creating a widget

fig <- ggplotly(map, tooltip = c("text"))

fig
